File: python_concepts/Synchronization/PrintFooBarAlternately/FooBar.py
from threading import *


class FooBar:
    def __init__(self, n):
        self.n = n
        self.fooSema: Semaphore = Semaphore(1)
        self.barSema: Semaphore = Semaphore(0)


    def foo(self, printFoo: 'Callable[[], None]') -> None:
        
        for i in range(self.n):
            self.fooSema.acquire()
            # printFoo() outputs "foo". Do not change or remove this line.
            printFoo()
            self.barSema.release()


    def bar(self, printBar: 'Callable[[], None]') -> None:
        
        for i in range(self.n):
            self.barSema.acquire()
            # printBar() outputs "bar". Do not change or remove this line.
            printBar()
            self.fooSema.release()

File: java_concepts/src/main/java/lld/Synchronization/Semaphores/PrintFooBarAlternately/FooBar.java
package lld.Synchronization.Semaphores.PrintFooBarAlternately;

import java.util.concurrent.Semaphore;

public class FooBar {
    private int n;
    Semaphore fooSema;
    Semaphore barSema;

    public FooBar(int n) {
        this.n = n;
        this.fooSema = new Semaphore(1);
        this.barSema = new Semaphore(0);
    }

    public void foo(Runnable printFoo) throws InterruptedException {
        
        for (int i = 0; i < n; i++) {
            fooSema.acquire();
        	// printFoo.run() outputs "foo". Do not change or remove this line.
        	printFoo.run();
            barSema.release();
        }
    }

    public void bar(Runnable printBar) throws InterruptedException {
        
        for (int i = 0; i < n; i++) {
            barSema.acquire();
            // printBar.run() outputs "bar". Do not change or remove this line.
        	printBar.run();
            fooSema.release();
        }
    }
}
